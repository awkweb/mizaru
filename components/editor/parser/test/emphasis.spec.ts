import Parser from '../'

describe('basic', () => {
    for (const syntax of ['*', '_']) {
        const doc = `${syntax}foo${syntax}`
        test(doc, () => {
            const { decorations, nodes } = Parser.parse(doc)
            expect(nodes).toEqual([
                {
                    from: 0,
                    to: 7,
                    type: 'paragraph',
                    marks: [{ from: 1, to: 6, type: 'emphasis' }],
                },
            ])
            expect(decorations).toEqual([
                { from: 1, to: 2, type: 'syntax' },
                { from: 5, to: 6, type: 'syntax' },
            ])
        })
    }
})

describe('with nested', () => {
    test('delete', () => {
        const doc = '*~~foo~~*'
        const { decorations, nodes } = Parser.parse(doc)
        expect(nodes).toEqual([
            {
                from: 0,
                to: 11,
                type: 'paragraph',
                marks: [
                    { from: 2, to: 9, type: 'delete' },
                    { from: 1, to: 10, type: 'emphasis' },
                ],
            },
        ])
        expect(decorations).toEqual([
            { from: 2, to: 4, type: 'syntax' },
            { from: 7, to: 9, type: 'syntax' },
            { from: 1, to: 2, type: 'syntax' },
            { from: 9, to: 10, type: 'syntax' },
        ])
    })

    test('inlineCode', () => {
        const doc = '*`foo`*'
        const { decorations, nodes } = Parser.parse(doc)
        expect(nodes).toEqual([
            {
                from: 0,
                to: 9,
                type: 'paragraph',
                marks: [
                    { from: 2, to: 7, type: 'inlineCode' },
                    { from: 1, to: 8, type: 'emphasis' },
                ],
            },
        ])
        expect(decorations).toEqual([
            { from: 2, to: 3, type: 'syntax' },
            { from: 6, to: 7, type: 'syntax' },
            { from: 1, to: 2, type: 'syntax' },
            { from: 7, to: 8, type: 'syntax' },
        ])
    })

    test('strong', () => {
        const doc = '*__foo__*'
        const { decorations, nodes } = Parser.parse(doc)
        expect(nodes).toEqual([
            {
                from: 0,
                to: 11,
                type: 'paragraph',
                marks: [
                    { from: 2, to: 9, type: 'strong' },
                    { from: 1, to: 10, type: 'emphasis' },
                ],
            },
        ])
        expect(decorations).toEqual([
            { from: 2, to: 4, type: 'syntax' },
            { from: 7, to: 9, type: 'syntax' },
            { from: 1, to: 2, type: 'syntax' },
            { from: 9, to: 10, type: 'syntax' },
        ])
    })
})
