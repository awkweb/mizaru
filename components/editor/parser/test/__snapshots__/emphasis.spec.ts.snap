// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`* backslash 1`] = `
Object {
  "counter": 16,
  "decorations": Array [
    Object {
      "from": 1,
      "to": 2,
      "type": "syntax",
    },
    Object {
      "from": 14,
      "to": 15,
      "type": "syntax",
    },
  ],
  "nodes": Array [
    Object {
      "from": 0,
      "marks": Array [
        Object {
          "from": 1,
          "to": 15,
          "type": "emphasis",
        },
      ],
      "to": 16,
      "type": "paragraph",
    },
  ],
}
`;

exports[`* escaped 1`] = `
Object {
  "counter": 16,
  "decorations": Array [],
  "nodes": Array [
    Object {
      "from": 0,
      "marks": Array [],
      "to": 16,
      "type": "paragraph",
    },
  ],
}
`;

exports[`* multiple words 1`] = `
Object {
  "counter": 15,
  "decorations": Array [
    Object {
      "from": 1,
      "to": 2,
      "type": "syntax",
    },
    Object {
      "from": 13,
      "to": 14,
      "type": "syntax",
    },
  ],
  "nodes": Array [
    Object {
      "from": 0,
      "marks": Array [
        Object {
          "from": 1,
          "to": 14,
          "type": "emphasis",
        },
      ],
      "to": 15,
      "type": "paragraph",
    },
  ],
}
`;

exports[`* single word 1`] = `
Object {
  "counter": 7,
  "decorations": Array [
    Object {
      "from": 1,
      "to": 2,
      "type": "syntax",
    },
    Object {
      "from": 5,
      "to": 6,
      "type": "syntax",
    },
  ],
  "nodes": Array [
    Object {
      "from": 0,
      "marks": Array [
        Object {
          "from": 1,
          "to": 6,
          "type": "emphasis",
        },
      ],
      "to": 7,
      "type": "paragraph",
    },
  ],
}
`;

exports[`* split across multiple lines 1`] = `
Object {
  "counter": 17,
  "decorations": Array [
    Object {
      "from": 1,
      "to": 2,
      "type": "syntax",
    },
    Object {
      "from": 15,
      "to": 16,
      "type": "syntax",
    },
  ],
  "nodes": Array [
    Object {
      "from": 0,
      "marks": Array [
        Object {
          "from": 1,
          "to": 16,
          "type": "emphasis",
        },
      ],
      "to": 17,
      "type": "paragraph",
    },
  ],
}
`;

exports[`* with nested delete 1`] = `
Object {
  "counter": 19,
  "decorations": Array [
    Object {
      "from": 6,
      "to": 8,
      "type": "syntax",
    },
    Object {
      "from": 11,
      "to": 13,
      "type": "syntax",
    },
    Object {
      "from": 1,
      "to": 2,
      "type": "syntax",
    },
    Object {
      "from": 17,
      "to": 18,
      "type": "syntax",
    },
  ],
  "nodes": Array [
    Object {
      "from": 0,
      "marks": Array [
        Object {
          "from": 6,
          "to": 13,
          "type": "delete",
        },
        Object {
          "from": 1,
          "to": 18,
          "type": "emphasis",
        },
      ],
      "to": 19,
      "type": "paragraph",
    },
  ],
}
`;

exports[`* with nested emphasis 1`] = `
Object {
  "counter": 17,
  "decorations": Array [
    Object {
      "from": 1,
      "to": 2,
      "type": "syntax",
    },
    Object {
      "from": 6,
      "to": 7,
      "type": "syntax",
    },
    Object {
      "from": 10,
      "to": 11,
      "type": "syntax",
    },
    Object {
      "from": 15,
      "to": 16,
      "type": "syntax",
    },
  ],
  "nodes": Array [
    Object {
      "from": 0,
      "marks": Array [
        Object {
          "from": 1,
          "to": 7,
          "type": "emphasis",
        },
        Object {
          "from": 10,
          "to": 16,
          "type": "emphasis",
        },
      ],
      "to": 17,
      "type": "paragraph",
    },
  ],
}
`;

exports[`* with nested inlineCode 1`] = `
Object {
  "counter": 17,
  "decorations": Array [
    Object {
      "from": 6,
      "to": 7,
      "type": "syntax",
    },
    Object {
      "from": 10,
      "to": 11,
      "type": "syntax",
    },
    Object {
      "from": 1,
      "to": 2,
      "type": "syntax",
    },
    Object {
      "from": 15,
      "to": 16,
      "type": "syntax",
    },
  ],
  "nodes": Array [
    Object {
      "from": 0,
      "marks": Array [
        Object {
          "from": 6,
          "to": 11,
          "type": "inlineCode",
        },
        Object {
          "from": 1,
          "to": 16,
          "type": "emphasis",
        },
      ],
      "to": 17,
      "type": "paragraph",
    },
  ],
}
`;

exports[`* with nested strong 1`] = `
Object {
  "counter": 19,
  "decorations": Array [
    Object {
      "from": 6,
      "to": 8,
      "type": "syntax",
    },
    Object {
      "from": 11,
      "to": 13,
      "type": "syntax",
    },
    Object {
      "from": 1,
      "to": 2,
      "type": "syntax",
    },
    Object {
      "from": 17,
      "to": 18,
      "type": "syntax",
    },
  ],
  "nodes": Array [
    Object {
      "from": 0,
      "marks": Array [
        Object {
          "from": 6,
          "to": 13,
          "type": "strong",
        },
        Object {
          "from": 1,
          "to": 18,
          "type": "emphasis",
        },
      ],
      "to": 19,
      "type": "paragraph",
    },
  ],
}
`;

exports[`_ backslash 1`] = `
Object {
  "counter": 16,
  "decorations": Array [
    Object {
      "from": 1,
      "to": 2,
      "type": "syntax",
    },
    Object {
      "from": 14,
      "to": 15,
      "type": "syntax",
    },
  ],
  "nodes": Array [
    Object {
      "from": 0,
      "marks": Array [
        Object {
          "from": 1,
          "to": 15,
          "type": "emphasis",
        },
      ],
      "to": 16,
      "type": "paragraph",
    },
  ],
}
`;

exports[`_ escaped 1`] = `
Object {
  "counter": 16,
  "decorations": Array [],
  "nodes": Array [
    Object {
      "from": 0,
      "marks": Array [],
      "to": 16,
      "type": "paragraph",
    },
  ],
}
`;

exports[`_ multiple words 1`] = `
Object {
  "counter": 15,
  "decorations": Array [
    Object {
      "from": 1,
      "to": 2,
      "type": "syntax",
    },
    Object {
      "from": 13,
      "to": 14,
      "type": "syntax",
    },
  ],
  "nodes": Array [
    Object {
      "from": 0,
      "marks": Array [
        Object {
          "from": 1,
          "to": 14,
          "type": "emphasis",
        },
      ],
      "to": 15,
      "type": "paragraph",
    },
  ],
}
`;

exports[`_ single word 1`] = `
Object {
  "counter": 7,
  "decorations": Array [
    Object {
      "from": 1,
      "to": 2,
      "type": "syntax",
    },
    Object {
      "from": 5,
      "to": 6,
      "type": "syntax",
    },
  ],
  "nodes": Array [
    Object {
      "from": 0,
      "marks": Array [
        Object {
          "from": 1,
          "to": 6,
          "type": "emphasis",
        },
      ],
      "to": 7,
      "type": "paragraph",
    },
  ],
}
`;

exports[`_ split across multiple lines 1`] = `
Object {
  "counter": 17,
  "decorations": Array [
    Object {
      "from": 1,
      "to": 2,
      "type": "syntax",
    },
    Object {
      "from": 15,
      "to": 16,
      "type": "syntax",
    },
  ],
  "nodes": Array [
    Object {
      "from": 0,
      "marks": Array [
        Object {
          "from": 1,
          "to": 16,
          "type": "emphasis",
        },
      ],
      "to": 17,
      "type": "paragraph",
    },
  ],
}
`;

exports[`_ with nested delete 1`] = `
Object {
  "counter": 19,
  "decorations": Array [
    Object {
      "from": 6,
      "to": 8,
      "type": "syntax",
    },
    Object {
      "from": 11,
      "to": 13,
      "type": "syntax",
    },
    Object {
      "from": 1,
      "to": 2,
      "type": "syntax",
    },
    Object {
      "from": 17,
      "to": 18,
      "type": "syntax",
    },
  ],
  "nodes": Array [
    Object {
      "from": 0,
      "marks": Array [
        Object {
          "from": 6,
          "to": 13,
          "type": "delete",
        },
        Object {
          "from": 1,
          "to": 18,
          "type": "emphasis",
        },
      ],
      "to": 19,
      "type": "paragraph",
    },
  ],
}
`;

exports[`_ with nested emphasis 1`] = `
Object {
  "counter": 17,
  "decorations": Array [
    Object {
      "from": 6,
      "to": 7,
      "type": "syntax",
    },
    Object {
      "from": 10,
      "to": 11,
      "type": "syntax",
    },
    Object {
      "from": 1,
      "to": 2,
      "type": "syntax",
    },
    Object {
      "from": 15,
      "to": 16,
      "type": "syntax",
    },
  ],
  "nodes": Array [
    Object {
      "from": 0,
      "marks": Array [
        Object {
          "from": 6,
          "to": 11,
          "type": "emphasis",
        },
        Object {
          "from": 1,
          "to": 16,
          "type": "emphasis",
        },
      ],
      "to": 17,
      "type": "paragraph",
    },
  ],
}
`;

exports[`_ with nested inlineCode 1`] = `
Object {
  "counter": 17,
  "decorations": Array [
    Object {
      "from": 6,
      "to": 7,
      "type": "syntax",
    },
    Object {
      "from": 10,
      "to": 11,
      "type": "syntax",
    },
    Object {
      "from": 1,
      "to": 2,
      "type": "syntax",
    },
    Object {
      "from": 15,
      "to": 16,
      "type": "syntax",
    },
  ],
  "nodes": Array [
    Object {
      "from": 0,
      "marks": Array [
        Object {
          "from": 6,
          "to": 11,
          "type": "inlineCode",
        },
        Object {
          "from": 1,
          "to": 16,
          "type": "emphasis",
        },
      ],
      "to": 17,
      "type": "paragraph",
    },
  ],
}
`;

exports[`_ with nested strong 1`] = `
Object {
  "counter": 19,
  "decorations": Array [
    Object {
      "from": 6,
      "to": 8,
      "type": "syntax",
    },
    Object {
      "from": 11,
      "to": 13,
      "type": "syntax",
    },
    Object {
      "from": 1,
      "to": 2,
      "type": "syntax",
    },
    Object {
      "from": 17,
      "to": 18,
      "type": "syntax",
    },
  ],
  "nodes": Array [
    Object {
      "from": 0,
      "marks": Array [
        Object {
          "from": 6,
          "to": 13,
          "type": "strong",
        },
        Object {
          "from": 1,
          "to": 18,
          "type": "emphasis",
        },
      ],
      "to": 19,
      "type": "paragraph",
    },
  ],
}
`;
